Makefile

```diff
@@ -5,9 +5,7 @@ VERSION=$(shell $(PYTHON) setup.py --version 2>/dev/null)
 PYTHON_DEVELOP_ARGS=$(shell if ($(PYTHON) setup.py develop --help 2>/dev/null | grep -q '\-\-user'); then echo "--user"; else echo ""; fi)
 DESTDIR=/
 AVOCADO_DIRNAME=$(shell basename ${PWD})
-AVOCADO_EXTERNAL_PLUGINS=$(filter-out ../$(AVOCADO_DIRNAME), $(shell find ../ -maxdepth 1 -mindepth 1 -type d))
```

Comment:
This already caused problems in my environment, and a solution is, indeed, welcome. My only concern is that it removes the Avocado capability of having external plugins without documenting or showing how it is handled now. This may impact external plugins we are not aware of.

Reply:
If we have external plugins then those external plugins should have their own targets.

This was used basically to clean and uninstall them. If they are external plugins then each plugin should take care of documenting this, IMO.

I'm having a hard time finding how this would impact them. Could you give me an example?

Reply:
`make link` was using it to make Avocado aware of external plugins. With this change, how can Avocado detect external plugins? I think it is a good feature from Avocado that we are silently removing.

Reply:
Unless I'm missing something, `make link` was installing "plugins" in "develop mode", and it was assuming that ***any*** external folders with a `Makefile` and/or `setup.py` was an Avocado plugin. This is a wrong assumption and dangerous.

If we have external plugins (outside Avocado folder), IMO is not Avocado's responsibility to call their `make targets`. Instead, we should properly list them as requirements if they are needed to run Avocado (in requirements.txt, setup.py, or any other way).

Imagine that I have 10 "valid" external plugins, but I'm installing Avocado for testing/using Avocado only, IMO makes no sense to install all those external plugins, since **there is no guarantee those plugins are compatible with the latest Avocado's version.** Why we should force developers to do that?

Avocado will not lose the ability to have external plugins, but those should be installed manually or as a "requirement" if needed.

Also having `link` and `developer` is confusing. Because `developer mode` is basically a link.

Bellow, I list the mapping that I'm imagining:

```
* install: python3 setup.py install
* develop: python3 setup.py install --develop (This is a replacement for link)
* uninstall: Would remove regardless of the mode that you installed
```

And we could also have an optional `--all` for saying that we would install the subprojects as well. But this refactoring would be bigger and part of other PR, if you guys agree on that.

But again, I might be missing something.

Reply:
> Unless I'm missing something, `make link` was installing "plugins" in "develop mode", and it was assuming that _**any**_ external folders with a `Makefile` and/or `setup.py` was an Avocado plugin. This is a wrong assumption and dangerous.

I totally agree with you!

> 
> If we have external plugins (outside Avocado folder), IMO is not Avocado's responsibility to call their `make targets`. Instead, we should properly list them as requirements if they are needed to run Avocado (in requirements.txt, setup.py, or any other way).

I also agree with you here that Avocado should have another way to detect external plugins.

> 
> Imagine that I have 10 "valid" external plugins, but I'm installing Avocado for testing/using Avocado only, IMO makes no sense to install all those external plugins, since **there is no guarantee those plugins are compatible with the latest Avocado's version.** Why we should force developers to do that?

We should not force, indeed.

> 
> Avocado will not lose the ability to have external plugins, but those should be installed manually or as a "requirement" if needed.

Okay, here is my point of concern, we are removing a feature that was supported by Avocado without a proper deprecation message and without an option for the existing plugins.

> 
> Also having `link` and `developer` is confusing. Because `developer mode` is basically a link.
> 
> Bellow, I list the mapping that I'm imagining:
> 
> ```
> * install: python3 setup.py install
> * develop: python3 setup.py install --develop (This is a replacement for link)
> * uninstall: Would remove regardless of the mode that you installed
> ```

That makes some sense but excludes the external plugins. I think the ability to support external plugins out of the box and without the need to interact with the Avocado project at all is a nice feature to have.

> 
> And we could also have an optional `--all` for saying that we would install the subprojects as well. But this refactoring would be bigger and part of other PR, if you guys agree on that.
> 
> But again, I might be missing something.

Reply:
> 
> > Avocado will not lose the ability to have external plugins, but those should be installed manually or as a "requirement" if needed.
> 
> Okay, here is my point of concern, we are removing a feature that was supported by Avocado without a proper deprecation message and without an option for the existing plugins.

We could add some notes on README and installation instructions, but I can't see where the deprecation message would fit here.

> > Also having `link` and `developer` is confusing. Because `developer mode` is basically a link.
> > Bellow, I list the mapping that I'm imagining:
> > ```
> > * install: python3 setup.py install
> > * develop: python3 setup.py install --develop (This is a replacement for link)
> > * uninstall: Would remove regardless of the mode that you installed
> > ```
> 
> That makes some sense but excludes the external plugins. I think the ability to support external plugins out of the box and without the need to interact with the Avocado project at all is a nice feature to have.

Sorry to insist here, but it is not clear to me "*why*" do you think it is a nice feature? Why should the Avocado installation install "third-party"/"optional"/"external" plugins? How we could install them if we don't have the guarantee that they work properly with our version?

As I said, the proper way IMO is adding the plugins as requirements if they are a requirement. Should we add all avocado plugins as requirement? IMO, not.

Reply:
> > > Imagine that I have 10 "valid" external plugins, but I'm installing Avocado for testing/using Avocado only, IMO makes no sense to install all those external plugins, since **there is no guarantee those plugins are compatible with the latest Avocado's version.** Why we should force developers to do that?
> 
> Well, this comes from the assumption that users _have_ to call `make link`. That's false. Users can call `make develop` or simply `setup.py develop` and no external plugin will be activated.

I'm trying to understand why "end-users", should call "develop/link" stuff?

 
> > We should not force, indeed.
> > > Avocado will not lose the ability to have external plugins, but those should be installed manually or as a "requirement" if needed.
> 
> Requirements won't cut as a substitute for `make link` (and for general development workflows) because you can't easily express that the `avocado-framework` requirement you want to use happens to be under a given directory. Sure, you can always run things manually (`cd $my-avocado-checkout; python setup.py develop; cd -; python setup.py develop), but that goes without saying.

I was trying to say that instead of getting those requirements from specific directories, we could adopt:

1. requirements-all-plugins.txt (or any other file describing the requirements) to get those via pip (either via PyPi or via Github URL), taking into consideration the proper versions match; or;
2. Your suggestion to create scripts to get those requirements.

I'm still trying to understand why we need a separated script if we could get those plugins via pip;

I'm not saying that we should not care about those " out side" plugins. Of course, we should. I'm just saying to set the boundaries here on this repo.

> > Okay, here is my point of concern, we are removing a feature that was supported by Avocado without a proper deprecation message and without an option for the existing plugins.
> > > Also having `link` and `developer` is confusing. Because `developer mode` is basically a link.
> > > Bellow, I list the mapping that I'm imagining:
> > > ```
> > > * install: python3 setup.py install
> > > * develop: python3 setup.py install --develop (This is a replacement for link)
> > > * uninstall: Would remove regardless of the mode that you installed
> > > ```
> > 
> > 
> > That makes some sense but excludes the external plugins. I think the ability to support external plugins out of the box and without the need to interact with the Avocado project at all is a nice feature to have.
> 
> Exactly, `make link` is a developer oriented target that is indeed a nice to have. I'd be fine with moving it out of the `Makefile`, to a place that regular users will have a harder time reaching. But, even having said that, all the other targets such as `make rpm`, `make rpm-release`, etc will have to be accounted for. Maybe even moving the Makefile itself elsewhere would do it. One extra piece of information that supports the distinction of regular users .vs. developers is that regular users will most of the time just do a `pip install avocado-framework`.

It is nice to have but not on setup.py or Makefile, because, IMO make develop (or make link) should take care of this project/repo only. For installing any "outside" plugin, we should adopt another solution. (vide my previous comment).

Reply:
> > > > Imagine that I have 10 "valid" external plugins, but I'm installing Avocado for testing/using Avocado only, IMO makes no sense to install all those external plugins, since **there is no guarantee those plugins are compatible with the latest Avocado's version.** Why we should force developers to do that?
> > 
> > 
> > Well, this comes from the assumption that users _have_ to call `make link`. That's false. Users can call `make develop` or simply `setup.py develop` and no external plugin will be activated.
> 
> I'm trying to understand why "end-users", should call "develop/link" stuff?
> 

If "users" are interacting with the source code, and want to install Avocado, they'd need to in order to run Avocado from that source code directory.

> > > We should not force, indeed.
> > > > Avocado will not lose the ability to have external plugins, but those should be installed manually or as a "requirement" if needed.
> > 
> > 
> > Requirements won't cut as a substitute for `make link` (and for general development workflows) because you can't easily express that the `avocado-framework` requirement you want to use happens to be under a given directory. Sure, you can always run things manually (`cd $my-avocado-checkout; python setup.py develop; cd -; python setup.py develop), but that goes without saying.
> 
> I was trying to say that instead of getting those requirements from specific directories, we could adopt:
> 
> 1. requirements-all-plugins.txt (or any other file describing the requirements) to get those via pip (either via PyPi or via Github URL), taking into consideration the proper versions match; or;

This doesn't work well if you're say, hacking on Avocado-VT.  This is a development oriented feature.  

> 2. Your suggestion to create scripts to get those requirements.
> 
> I'm still trying to understand why we need a separated script if we could get those plugins via pip;
> 

`make link` is not intended for fetching commited/merged/released plugins.  It's a *development* time helper.

> I'm not saying that we should not care about those " out side" plugins. Of course, we should. I'm just saying to set the boundaries here on this repo.
> 
> > > Okay, here is my point of concern, we are removing a feature that was supported by Avocado without a proper deprecation message and without an option for the existing plugins.
> > > > Also having `link` and `developer` is confusing. Because `developer mode` is basically a link.
> > > > Bellow, I list the mapping that I'm imagining:
> > > > ```
> > > > * install: python3 setup.py install
> > > > * develop: python3 setup.py install --develop (This is a replacement for link)
> > > > * uninstall: Would remove regardless of the mode that you installed
> > > > ```
> > > 
> > > 
> > > That makes some sense but excludes the external plugins. I think the ability to support external plugins out of the box and without the need to interact with the Avocado project at all is a nice feature to have.
> > 
> > 
> > Exactly, `make link` is a developer oriented target that is indeed a nice to have. I'd be fine with moving it out of the `Makefile`, to a place that regular users will have a harder time reaching. But, even having said that, all the other targets such as `make rpm`, `make rpm-release`, etc will have to be accounted for. Maybe even moving the Makefile itself elsewhere would do it. One extra piece of information that supports the distinction of regular users .vs. developers is that regular users will most of the time just do a `pip install avocado-framework`.
> 
> It is nice to have but not on setup.py or Makefile, because, IMO make develop (or make link) should take care of this project/repo only. For installing any "outside" plugin, we should adopt another solution. (vide my previous comment).

If it makes you happy, like I said previously, we can simply move the `Makefile` elsewhere (`contrib/devel/`).

Reply:
> > > We should not force, indeed.
> > > > Avocado will not lose the ability to have external plugins, but those should be installed manually or as a "requirement" if needed.
> > 
> > 
> > Requirements won't cut as a substitute for `make link` (and for general development workflows) because you can't easily express that the `avocado-framework` requirement you want to use happens to be under a given directory. Sure, you can always run things manually (`cd $my-avocado-checkout; python setup.py develop; cd -; python setup.py develop), but that goes without saying.
> 
> I was trying to say that instead of getting those requirements from specific directories, we could adopt:
> 
> 1. requirements-all-plugins.txt (or any other file describing the requirements) to get those via pip (either via PyPi or via Github URL), taking into consideration the proper versions match; or;

> This doesn't work well if you're say, hacking on Avocado-VT.  This is a development oriented feature.  

Sorry for insisting on this, but If someone is hacking on Avocado-VT then a "python setup.py develop" on avocado-vt wouldn't be enough?

> 2. Your suggestion to create scripts to get those requirements.
> 
> I'm still trying to understand why we need a separated script if we could get those plugins via pip;
> 

> `make link` is not intended for fetching commited/merged/released plugins.  It's a *development* time helper. 

I understand that and I think that we are talking about the same thing.

> I'm not saying that we should not care about those " out side" plugins. Of course, we should. I'm just saying to set the boundaries here on this repo.
> 
> > > Okay, here is my point of concern, we are removing a feature that was supported by Avocado without a proper deprecation message and without an option for the existing plugins.
> > > > Also having `link` and `developer` is confusing. Because `developer mode` is basically a link.
> > > > Bellow, I list the mapping that I'm imagining:
> > > > ```
> > > > * install: python3 setup.py install
> > > > * develop: python3 setup.py install --develop (This is a replacement for link)
> > > > * uninstall: Would remove regardless of the mode that you installed
> > > > ```
> > > 
> > > 
> > > That makes some sense but excludes the external plugins. I think the ability to support external plugins out of the box and without the need to interact with the Avocado project at all is a nice feature to have.
> > 
> > 
> > Exactly, `make link` is a developer oriented target that is indeed a nice to have. I'd be fine with moving it out of the `Makefile`, to a place that regular users will have a harder time reaching. But, even having said that, all the other targets such as `make rpm`, `make rpm-release`, etc will have to be accounted for. Maybe even moving the Makefile itself elsewhere would do it. One extra piece of information that supports the distinction of regular users .vs. developers is that regular users will most of the time just do a `pip install avocado-framework`.
> 
> It is nice to have but not on setup.py or Makefile, because, IMO make develop (or make link) should take care of this project/repo only. For installing any "outside" plugin, we should adopt another solution. (vide my previous comment).

> If it makes you happy, like I said previously, we can simply move the `Makefile` elsewhere (`contrib/devel/`).

It is not just a matter of moving this out of my yes. but also I would like to make sure that we are not just moving the dust under the carpet by moving to another place. I'm trying to create a good experience for new contributors.

Maybe it is time to start a new discussion (maybe this thread is getting bigger already) about our possible Make/setup targets.

Reply:
> > > > We should not force, indeed.
> > > > > Avocado will not lose the ability to have external plugins, but those should be installed manually or as a "requirement" if needed.
> > > 
> > > 
> > > Requirements won't cut as a substitute for `make link` (and for general development workflows) because you can't easily express that the `avocado-framework` requirement you want to use happens to be under a given directory. Sure, you can always run things manually (`cd $my-avocado-checkout; python setup.py develop; cd -; python setup.py develop), but that goes without saying.
> > 
> > 
> > I was trying to say that instead of getting those requirements from specific directories, we could adopt:
> > 
> > 1. requirements-all-plugins.txt (or any other file describing the requirements) to get those via pip (either via PyPi or via Github URL), taking into consideration the proper versions match; or;
> 
> > This doesn't work well if you're say, hacking on Avocado-VT.  This is a development oriented feature.
> 
> Sorry for insisting on this, but If someone is hacking on Avocado-VT then a "python setup.py develop" on avocado-vt wouldn't be enough?
> 

Like I said before, I'd be more like: `cd $my-avocado-checkout; python setup.py develop; cd -; python setup.py develop`. That is without taking into consideration the clean up.

> > 1. Your suggestion to create scripts to get those requirements.
> > 
> > I'm still trying to understand why we need a separated script if we could get those plugins via pip;
> 
> > `make link` is not intended for fetching commited/merged/released plugins.  It's a _development_ time helper.
> 
> I understand that and I think that we are talking about the same thing.
> 

OK, so I think we can the "fulfill by requirements" idea out of the way?  Do we agree that they're not optimal for this problem?

> > I'm not saying that we should not care about those " out side" plugins. Of course, we should. I'm just saying to set the boundaries here on this repo.
> > > > Okay, here is my point of concern, we are removing a feature that was supported by Avocado without a proper deprecation message and without an option for the existing plugins.
> > > > > Also having `link` and `developer` is confusing. Because `developer mode` is basically a link.
> > > > > Bellow, I list the mapping that I'm imagining:
> > > > > ```
> > > > > * install: python3 setup.py install
> > > > > * develop: python3 setup.py install --develop (This is a replacement for link)
> > > > > * uninstall: Would remove regardless of the mode that you installed
> > > > > ```
> > > > 
> > > > 
> > > > That makes some sense but excludes the external plugins. I think the ability to support external plugins out of the box and without the need to interact with the Avocado project at all is a nice feature to have.
> > > 
> > > 
> > > Exactly, `make link` is a developer oriented target that is indeed a nice to have. I'd be fine with moving it out of the `Makefile`, to a place that regular users will have a harder time reaching. But, even having said that, all the other targets such as `make rpm`, `make rpm-release`, etc will have to be accounted for. Maybe even moving the Makefile itself elsewhere would do it. One extra piece of information that supports the distinction of regular users .vs. developers is that regular users will most of the time just do a `pip install avocado-framework`.
> > 
> > 
> > It is nice to have but not on setup.py or Makefile, because, IMO make develop (or make link) should take care of this project/repo only. For installing any "outside" plugin, we should adopt another solution. (vide my previous comment).
> 
> > If it makes you happy, like I said previously, we can simply move the `Makefile` elsewhere (`contrib/devel/`).
> 
> It is not just a matter of moving this out of my yes. but also I would like to make sure that we are not just moving the dust under the carpet by moving to another place. I'm trying to create a good experience for new contributors.
> 

I don't consider something I rely on every time I'm hacking on Avocado and external plugins to be the equivalent of dust.  The set of actions to setup Avocado and Avocado-VT (or other external plugins) is useful, and I see no reason to scrap and ask every Avocado+Avocado-VT developer to come up with its own.  Do you agree with that?

If you do, then the question really becomes how to mitigate/eliminate the annoyances, without loosing the value.  Even moving that to Avocado-VT *could* be done, but the downside is that it'd have to replicated on every external plugin.  

> Maybe it is time to start a new discussion (maybe this thread is getting bigger already) about our possible Make/setup targets.

I think we're close to reaching convergence here... let's see how the next iteration goes ;).

Reply:
> > Sorry for insisting on this, but If someone is hacking on Avocado-VT then a "python setup.py develop" on avocado-vt wouldn't be enough?
> 
> Like I said before, I'd be more like: `cd $my-avocado-checkout; python setup.py develop; cd -; python setup.py develop`. That is without taking into consideration the clean up.
> 
> > > 1. Your suggestion to create scripts to get those requirements.
> > > 
> > > I'm still trying to understand why we need a separated script if we could get those plugins via pip;
> > 
> > 
> > > `make link` is not intended for fetching commited/merged/released plugins.  It's a _development_ time helper.
> > 
> > 
> > I understand that and I think that we are talking about the same thing.
> 
> OK, so I think we can the "fulfill by requirements" idea out of the way? Do we agree that they're not optimal for this problem?

I promise that I'm trying hard to find the problem with this approach:

```bash
$ cat setup-developer-env.txt
-e git://github.com/avocado-framework/avocado-vt.git#egg=avocado-vt
# any other plugin here.

$ pip install -r setup-developer-env.txt
```
This will clone the repo into your virtualenv or local pip folder, and install it in developer mode. Letting the directory "ready" for Avocado-VT hacking. Very similar to what I'm imagining about what the "new script" that you mention would do. 


> > It is not just a matter of moving this out of my yes. but also I would like to make sure that we are not just moving the dust under the carpet by moving to another place. I'm trying to create a good experience for new contributors.
> 
> I don't consider something I rely on every time I'm hacking on Avocado and external plugins to be the equivalent of dust. The set of actions to setup Avocado and Avocado-VT (or other external plugins) is useful, and I see no reason to scrap and ask every Avocado+Avocado-VT developer to come up with its own. Do you agree with that?

Dust it was a bad choice. And yes, I completely agree with that.

Reply:
> > > Sorry for insisting on this, but If someone is hacking on Avocado-VT then a "python setup.py develop" on avocado-vt wouldn't be enough?
> > 
> > 
> > Like I said before, I'd be more like: `cd $my-avocado-checkout; python setup.py develop; cd -; python setup.py develop`. That is without taking into consideration the clean up.
> > > > 1. Your suggestion to create scripts to get those requirements.
> > > > 
> > > > I'm still trying to understand why we need a separated script if we could get those plugins via pip;
> > > 
> > > 
> > > > `make link` is not intended for fetching commited/merged/released plugins.  It's a _development_ time helper.
> > > 
> > > 
> > > I understand that and I think that we are talking about the same thing.
> > 
> > 
> > OK, so I think we can the "fulfill by requirements" idea out of the way? Do we agree that they're not optimal for this problem?
> 
> I promise that I'm trying hard to find the problem with this approach:
> 
> ```shell
> $ cat setup-developer-env.txt
> -e git://github.com/avocado-framework/avocado-vt.git#egg=avocado-vt
> # any other plugin here.
> 
> $ pip install -r setup-developer-env.txt
> ```
> 
> This will clone the repo into your virtualenv or local pip folder, and install it in developer mode. Letting the directory "ready" for Avocado-VT hacking. Very similar to what I'm imagining about what the "new script" that you mention would do.
> 

Do you checkout the project everytime or intend to hack on it? Do you hack on a GIT checkout made by pip?  And then what after that?  Do you configure remotes on the temporary GIT checkout and then push to your remote repo?  Then you destroy the venv with that checkou? That sounds painful to me.

My workflow is:

1) cd ~/src/avocado-development-directory/avocado-vt
1.1) git checkout -b topic_branch
2) cd ../avocado
2.1) git checkout master|82lts|69lts
2.2) make link
3) cd ../avocado-vt
3.1) hack; hack; hack
3.2) git commit -m hacked_the_heck_out_of_it
3.3) git push

Profit.  That can be done in venv *if wanted*, it makes no difference.

Basically, I can't see how using pip to checkout a repo I'll work on everytime is a good thing.  Sometimes I'll rebase a topic branch onto another WIP branch... I can't really imagine how to do that depending on pip.

Let me know if I failed to understand your proposal.

Cheers.  

> > > It is not just a matter of moving this out of my yes. but also I would like to make sure that we are not just moving the dust under the carpet by moving to another place. I'm trying to create a good experience for new contributors.
> > 
> > 
> > I don't consider something I rely on every time I'm hacking on Avocado and external plugins to be the equivalent of dust. The set of actions to setup Avocado and Avocado-VT (or other external plugins) is useful, and I see no reason to scrap and ask every Avocado+Avocado-VT developer to come up with its own. Do you agree with that?
> 
> Dust it was a bad choice. And yes, I completely agree with that.

Reply:
I have a very similar workflow.

The file+pip was not intended to be executed *every* time that we need to hack. Just for the first time, it is a bootstrap. Branches could be created after that, yes. And yes, there is the drawback of overriding that could be mitigated somehow.

Maybe things will get clear to me when we decide on what we should in behalf of other developers to help the bootstrap for them. What do you imagine that would be part of the `devel-bootstrap.sh|py` script?

(btw, thanks for your patience again!)

Reply:
> I have a very similar workflow.
> 

OK, that's good.  Can you then think of how you can replace/improve the current `make link` to fit into that workflow?  It can be a patch if you feel it's hard to explain it.

> The file+pip was not intended to be executed _every_ time that we need to hack. Just for the first time, it is a bootstrap. Branches could be created after that, yes. And yes, there is the drawback of overriding that could be mitigated somehow.
> 

Right, but I feel like a *new* repo is created and cloned with that step, right?  It does *not* reuse your existing repo (which may have configuration, other branches, `rerere` metadata, etc), correct?

> Maybe things will get clear to me when we decide on what we should do to help the bootstrap of other developers. What do you imagine that would be part of the `devel-bootstrap.sh|py` script?
> 

IMO it would be similar to what is done Today with `make link`.  If we want to *add* a use case where a new script sets up Avocado + Avocado-VT for development (inside a venv or not), fine.  But I don't think we should *remove* the current use case.

What I would ask of you is: code an improvement that removes the currently dangerous aspect that allows any developer to do a `make link` and will result in `make clean` and similar stuff being executed on all directories at the same level, without letting users that really want to keep doing that achieve it.  It could be done by:

1) moving the `Makefile` or some targets elsewhere (would require `make -C contrib/devel link`)
2) changing `make link` so that it would look at directories elsewhere if `AVOCADO_DEVEL_LOOK_AT_TOP_DIRECTORIES_FOR_PLUGINS` is set
3) writing a `devel-bootstrap.sh|py` 

Or any other solution you come up with.

> (btw, thanks for your patience again!)

No need to thank me... this is you trying to improve things, and I'm very much grateful for it!

Reply:
> > I have a very similar workflow.
> 
> OK, that's good. Can you then think of how you can replace/improve the current `make link` to fit into that workflow? It can be a patch if you feel it's hard to explain it.

First, let me say something to avoid confusion: "link" and "develop" targets IMO are very similar/confusing ("develop" is doing links as well, and "link" is doing "develop" as well) and because of that I would choose only "develop" for Makefile/setup target here.  I would choose "develop" for installing avocado + plugins in develop mode. And 'develop-external' (or something similar) for installing external plugins in develop mode too.

Now, the main reason that I don't think we should "guarantee" installation in third-party plugins, is because external plugins are developed by third parties and we don't have control over their build system. They could be using Makefile with different targets, and they could be using something else different from setup.py.

If that makes sense and IIUC, then the only "external plugins" we have "control" are the ones that we have in our GH namespace: `avocado-vt`, `avocado-virt`, `cit-ng`, and `avocado-ec2`. And I'm not sure if today we are using these plugins, except avocado-vt, and all other optional plugins that we intend to move outside `avocado` repo soon.

So, unless I'm missing something, we are talking about one script (or target) to install `avocado-vt` in develop mode. But ok, I will assume that this will be used by other plugins that will come up soon (like the optional-plugins). And we could also assume that those plugins will have a standard `setup.py` target to install in `develop` mode (`python3 setup.py develop`).

So, this "develop-external" script would be only a loop executing this "develop" target in all optional plugins that we have. Developers should point in the system where are those projects locally. Assuming any specific path is enforcing a developer environment, what IMO is not necessary, since it will be just a matter of running a standard `python3 setup.py develop` on each repo. 
 
Let me know if those assumptions make sense to you.

> > The file+pip was not intended to be executed _every_ time that we need to hack. Just for the first time, it is a bootstrap. Branches could be created after that, yes. And yes, there is the drawback of overriding that could be mitigated somehow.
> 
> Right, but I feel like a _new_ repo is created and cloned with that step, right? It does _not_ reuse your existing repo (which may have configuration, other branches, `rerere` metadata, etc), correct?

The example that I gave was just to illustrate, of course, we could think about something that does not override the repo. Let's forget the usage of pip for this for now. (let's focus on the content of the bootstrap script).

> 
> > Maybe things will get clear to me when we decide on what we should do to help the bootstrap of other developers. What do you imagine that would be part of the `devel-bootstrap.sh|py` script?
> 
> IMO it would be similar to what is done Today with `make link`. If we want to _add_ a use case where a new script sets up Avocado + Avocado-VT for development (inside a venv or not), fine. But I don't think we should _remove_ the current use case.
> 
> What I would ask of you is: code an improvement that removes the currently dangerous aspect that allows any developer to do a `make link` and will result in `make clean` and similar stuff being executed on all directories at the same level, without letting users that really want to keep doing that achieve it. It could be done by:
> 
>     1. moving the `Makefile` or some targets elsewhere (would require `make -C contrib/devel link`)

What I don't like about this approach: We will have several places where the targets are (more than we have today): `Makefile`, `setup.py` and `contrib/devel`. I'm trying to reduce the number of places. And to move the entire Makefile to contrib we need to implement all the current/valid targets into setup.py (That I have intentions to do, but with baby steps).

>     2. changing `make link` so that it would look at directories elsewhere if `AVOCADO_DEVEL_LOOK_AT_TOP_DIRECTORIES_FOR_PLUGINS` is set
> 
>     3. writing a `devel-bootstrap.sh|py`

If you don't mind, I will let this for the future since it was just a brainstorm, we don't have this right now.

So, If my comments make sense, today this "link/develop-external" would be used only for one plugin (avocado-vt) and IMO it is over need.

But I think that I got the overall idea, I just would like to make sure that I have the same understanding as you do have here and I will change this PR, but I will also rename the targets if you don't mind.

Reply:
@clebergnu I just force-pushed here with an alternative/suggestion. But it would be nice to read your thoughts about my previous comments too.

Reply:
Yes, I'm totally fine the solution you proposed.